akka-http-core/src/main/scala/akka/http/impl/engine/client/PoolSlot.scala

```diff
@@ -89,187 +76,163 @@ private object PoolSlot {
-        request(n - remainingRequested)
-        connInport ! OnNext(firstRequest.request)
-        context.become(running(connInport, connOutport))
-
-      case FromConnection(Cancel)     ⇒ if (!isActive) { cancel(); shutdown() } // else ignore and wait for accompanying OnComplete or OnError
-      case FromConnection(OnComplete) ⇒ handleDisconnect(sender(), None, Some(firstRequest))
-      case FromConnection(OnError(e)) ⇒ handleDisconnect(sender(), Some(e), Some(firstRequest))
-      case FromConnection(OnNext(x))  ⇒ throw new IllegalStateException("Unexpected HttpResponse: " + x)
-    }
+          }
+        }
+      }
 
-    def running(connInport: ActorRef, connOutport: ActorRef): Receive = {
-      case ev @ (Request(_) | Cancel)     ⇒ connOutport ! ev
-      case ev @ (OnComplete | OnError(_)) ⇒ connInport ! ev
-      case OnNext(DispatchCommand(rc: RequestContext)) ⇒
-        inflightRequests = inflightRequests.enqueue(rc)
-        connInport ! OnNext(rc.request)
-
-      case FromConnection(Request(n)) ⇒ request(n)
-      case FromConnection(Cancel)     ⇒ if (!isActive) { cancel(); shutdown() } // else ignore and wait for accompanying OnComplete or OnError
-
-      case FromConnection(OnNext(response: HttpResponse)) ⇒
-        val requestContext = inflightRequests.head
-        inflightRequests = inflightRequests.tail
-        val (entity, whenCompleted) = HttpEntity.captureTermination(response.entity)
-        val delivery = ResponseDelivery(ResponseContext(requestContext, Success(response withEntity entity)))
-        import fm.executionContext
-        val requestCompleted = SlotEvent.RequestCompletedFuture(whenCompleted.map(_ ⇒ SlotEvent.RequestCompleted(slotIx)))
-        onNext(delivery :: requestCompleted :: Nil)
-
-      case FromConnection(OnComplete) ⇒ handleDisconnect(sender(), None)
-      case FromConnection(OnError(e)) ⇒ handleDisconnect(sender(), Some(e))
-    }
+      private lazy val connectionInFlowHandler = new InHandler {
+
+        // a new element is available on connectionFlowSink Inlet - that is a HttpResponse is being returned
+        override def onPush(): Unit = {
+          log.debug("BERN-{}: connectionFlow: onPush", slotIx)
+          // consume a HttpResponse from the connectonFlow
+
+          val response: HttpResponse = connectionFlowSink.grab()
+
+          log.debug("BERN-{}: connectionFlow: onPush {} {}", slotIx, response)
+          val requestContext = inflightRequests.head
+          inflightRequests = inflightRequests.tail
+
+          val (entity, whenCompleted) = HttpEntity.captureTermination(response.entity)
+          val delivery = ResponseDelivery(ResponseContext(requestContext, Success(response withEntity entity)))
+          import fm.executionContext
+          val requestCompleted = SlotEvent.RequestCompletedFuture(whenCompleted.map(_ ⇒ SlotEvent.RequestCompleted(slotIx)))
+          push(out, delivery :: requestCompleted :: Nil)
```

Comment:
Do this instead:

```scala
emitMultiple(out, delivery :: requestCompleted :: Nil)
```

Then we can make the output type not a `List[]` anymore, which allows us to remove:

```scala
.mapConcat(ConstantFun.scalaIdentityFunction))
```

So win-win